### PR TITLE
[DF] Introduced Standard Deviation Action

### PR DESCRIPTION
**Pull request by imaxoi**

### CODE REVIEW

**Comment by bluehood** [comment]
@phsft-bot build

File: tree/dataframe/test/dataframe_simple.cxx [hunk 1]
```diff
@@ -513,6 +513,26 @@ TEST_P(RDFSimpleTests, BookCustomAction)
    EXPECT_EQ(*maxSlot, nWorkers-1);
 }
 
+TEST(RDFSimpleTests, StandardDeviation)
+{
+   RDataFrame rd1(8);
+   int i(0);
+   auto stdDev = rd1.Define("b1",
+                            [&i]() {
+                               ++i;
```
Comment (bluehood) [inline tree/dataframe/test/dataframe_simple.cxx]: this is not thread-safe

File: tree/dataframe/test/dataframe_simple.cxx [hunk 2]
```diff
@@ -513,6 +513,26 @@ TEST_P(RDFSimpleTests, BookCustomAction)
...
+                               return i;
+                            })
+                    .StdDeviation<int>("b1");
```
Comment (bluehood) [inline tree/dataframe/test/dataframe_simple.cxx]: please also add a test for `StdDeviation("b1")`, i.e. no template parameter and therefore jitting

File: tree/dataframe/src/RDFActionHelpers.cxx [hunk 3]
```diff
@@ -163,6 +163,60 @@ template void MeanHelper::Exec(unsigned int, const std::vector<char> &);
...
+   auto count = ++fCounts[slot];
+   auto delta = v - fMeans[slot];
+   auto mean = fMeans[slot] + delta / count;
+   auto delta2 = v - mean;
+   auto distance = fDistancesfromMean[slot] + delta * delta2;
+
+   fCounts[slot] = count;
+   fMeans[slot] = mean;
+   fDistancesfromMean[slot] = distance;
+}
```
Comment (dpiparo) [inline tree/dataframe/src/RDFActionHelpers.cxx]: Are we sure of this algorithm? Is it numerically stable? Can we avoid the division? Can we make it incremental?

File: tree/dataframe/src/RDFActionHelpers.cxx [hunk 4]
```diff
@@ -163,6 +163,60 @@ template void MeanHelper::Exec(unsigned int, const std::vector<char> &);
...
+
+   double variance = 0;
+   for (unsigned int i = 0; i < fNSlots; ++i) {
+      if (fCounts[i] == 0) {
+         continue;
+      }
+      variance += fCounts[i] * (fDistancesfromMean[i] / fCounts[i] + std::pow((fMeans[i] - overallMean), 2));
+   }
+
+   variance = variance / totalElements;
```
Comment (stwunsch) [inline tree/dataframe/src/RDFActionHelpers.cxx]: You put here the biased estimation of the standard deviation. Do you want the unbiased one? -> `std = sqrt( 1 / (N - 1) * ... )`

https://en.wikipedia.org/wiki/Standard_deviation

https://en.wikipedia.org/wiki/Unbiased_estimation_of_standard_deviation

Comment (stwunsch) [reply tree/dataframe/src/RDFActionHelpers.cxx]: Does matter only for very small number of samples.

File: tree/dataframe/inc/ROOT/RDFActionHelpers.hxx [hunk 5]
```diff
@@ -640,6 +640,43 @@ extern template void MeanHelper::Exec(unsigned int, const std::vector<char> &);
...
+
+   void Finalize();
+};
+
+extern template void StdDevHelper::Exec(unsigned int, const std::vector<float> &);
+extern template void StdDevHelper::Exec(unsigned int, const std::vector<double> &);
+extern template void StdDevHelper::Exec(unsigned int, const std::vector<char> &);
+extern template void StdDevHelper::Exec(unsigned int, const std::vector<int> &);
+extern template void StdDevHelper::Exec(unsigned int, const std::vector<unsigned int> &);
+
```
Comment (dpiparo) [inline tree/dataframe/inc/ROOT/RDFActionHelpers.hxx]: 👍

File: tree/dataframe/src/RDFActionHelpers.cxx (hunk 4 above)
Comment (stwunsch) [reply tree/dataframe/src/RDFActionHelpers.cxx]: `RMS` is not a proper name for this function....

```
////////////////////////////////////////////////////////////////////////////////
/// Return the Standard Deviation of an array defined by the iterators.
/// Note that this function returns the sigma(standard deviation) and
/// not the root mean square of the array.
///
/// Use the two pass algorithm, which is slower (! a factor of 2) but much more
/// precise.  Since we have a vector the 2 pass algorithm is still faster than the
/// Welford algorithm. (See also ROOT-5545)
template <typename Iterator>
Double_t TMath::RMS(Iterator first, Iterator last)
{

   Double_t n = 0;

   Double_t tot = 0;
   Double_t mean = TMath::Mean(first,last);
   while ( first != last ) {
      Double_t x = Double_t(*first);
      tot += (x - mean)*(x - mean);
      ++first;
      ++n;
   }
   Double_t rms = (n > 1) ? TMath::Sqrt(tot/(n-1)) : 0.0;
   return rms;
}
```

File: tree/dataframe/test/dataframe_simple.cxx (hunk 1 above)
Comment (imaxoi) [reply tree/dataframe/test/dataframe_simple.cxx]: Fixed in 3e06584fa051988ffa8e28ddd6619f9c2ba885ce

File: tree/dataframe/src/RDFActionHelpers.cxx (hunk 3 above)
Comment (imaxoi) [reply tree/dataframe/src/RDFActionHelpers.cxx]: It's the Welford's algorithm. I think it is already incremental, as at the end of each step you know the partial result for the samples up to that point, and by the end you merge everything. I am going to investigate about the division.

>  John D. Cook found the accuracy of this method comparable to the accuracy of the two-pass method derived directly from the definition [cit.](http://jonisalonen.com/2013/deriving-welfords-method-for-computing-variance/)

> This algorithm is much less prone to loss of precision due to catastrophic cancellation, but might not be as efficient because of the division operation inside the loop. [cit.](https://en.wikipedia.org/wiki/Algorithms_for_calculating_variance#Online_algorithm)

File: tree/dataframe/src/RDFActionHelpers.cxx (hunk 4 above)
Comment (imaxoi) [reply tree/dataframe/src/RDFActionHelpers.cxx]: Thanks, I changed the formula to the unbiased one 4d22b9610974e5856d73d856626a6e04a1095008

File: tree/dataframe/src/RDFActionHelpers.cxx [hunk 6]
```diff
@@ -204,10 +206,11 @@ void StdDevHelper::Finalize()
       if (fCounts[i] == 0) {
          continue;
       }
-      variance += fCounts[i] * (fDistancesfromMean[i] / fCounts[i] + std::pow((fMeans[i] - overallMean), 2));
+      auto setVariance = fDistancesfromMean[i] / (fCounts[i] - 1);
+      variance += (fCounts[i] - 1) * (setVariance + std::pow((fMeans[i] - overallMean), 2));
    }
 
-   variance = variance / totalElements;
+   variance = variance / (totalElements - totalSets);
```
Comment (stwunsch) [inline tree/dataframe/src/RDFActionHelpers.cxx]: Is it correct? As it is Welford's algorithm (did not fully understand this in my first comment), I'm not sure whether it's biased or unbiased by default (or whether `N -> N - 1` does the trick).

File: tree/dataframe/src/RDFActionHelpers.cxx [hunk 7]
```diff
@@ -204,10 +206,11 @@ void StdDevHelper::Finalize()
       if (fCounts[i] == 0) {
          continue;
       }
-      variance += fCounts[i] * (fDistancesfromMean[i] / fCounts[i] + std::pow((fMeans[i] - overallMean), 2));
+      auto setVariance = fDistancesfromMean[i] / (fCounts[i] - 1);
```
Comment (imaxoi) [inline tree/dataframe/src/RDFActionHelpers.cxx]: @stwunsch 
Here the Welford's algorithm produces an unbiased variance for a subset, as in
`(mean, variance, sampleVariance) = (mean, M2/count, M2/(count - 1)) ` from [wiki](https://en.wikipedia.org/wiki/Algorithms_for_calculating_variance#Online_algorithm).

File: tree/dataframe/src/RDFActionHelpers.cxx (hunk 6 above)
Comment (imaxoi) [reply tree/dataframe/src/RDFActionHelpers.cxx]: @stwunsch Here there `totalSets` non-empty set, each one with a unbiased variance, so the unpooled one can be calculated.

**Comment by bluehood** [comment]
@imaxoi are the cases of zero and one entries handled correctly or e.g. do they divide by zero?

File: tree/dataframe/inc/ROOT/RDFInterface.hxx [hunk 8]
```diff
@@ -1213,6 +1213,24 @@ public:
...
+   /// \tparam T The type of the branch/column.
+   /// \param[in] columnName The name of the branch/column to be treated.
+   ///
+   /// If T is not specified, RDataFrame will infer it from the data and just-in-time compile the correct
+   /// template specialization of this method.
+   ///
+   /// This action is *lazy*: upon invocation of this method the calculation is
+   /// booked but not executed. See RResultPtr documentation.
+   template <typename T = RDFDetail::TInferType>
+   RResultPtr<double> StdDeviation(std::string_view columnName = "")
```
Comment (bluehood) [inline tree/dataframe/inc/ROOT/RDFInterface.hxx]: @dpiparo should we just call it `StdDev`? `StdDeviation` seems a bit verbose for something everybody will know.

File: tree/dataframe/src/RDFActionHelpers.cxx (hunk 3 above)
Comment (imaxoi) [reply tree/dataframe/src/RDFActionHelpers.cxx]: I followed @stwunsch suggestion and tried to run the Welford's algorithm against the deviation evaluated using its definition. Even for few samples (that should be the most critical part) the difference is in the order of ^-16. I suppose the algorithm is good.

Comment (bluehood) [reply tree/dataframe/src/RDFActionHelpers.cxx]: @imaxoi that's great. few samples are not necessarily the critical case, many samples might cause problems due to numerical instabilities in the summations.

in any case, since you already have the code, can you make it a gtest?

**Comment by imaxoi** [comment]
@bluehood Renamed to StdDev and added more tests for corner cases

File: tree/dataframe/src/RDFActionHelpers.cxx (hunk 3 above)
Comment (imaxoi) [reply tree/dataframe/src/RDFActionHelpers.cxx]: Ok 7e0b87e641eacb69695083fd90c52dd65cdd2b45

**Comment by dpiparo** [comment]
@imaxoi, can the action helper deal with columns holding collections, i.e. iterating over all elements? We do this for example for histogramming, it's just a matter of using a sfinae.

**Comment by imaxoi** [comment]
@dpiparo Added test in  f8fccd52ffcfc059aa5a664404e04c77d15abc99

File: tree/dataframe/test/dataframe_simple.cxx [hunk 9]
```diff
@@ -513,6 +514,113 @@ TEST_P(RDFSimpleTests, BookCustomAction)
...
+      return *d.DefineSlotEntry("x",
+                                [&samples](unsigned int slot, ULong64_t entry) {
+                                   (void)slot;
+                                   return samples[entry];
+                                })
+                 .StdDev("x");
+   }
+};
+
+TEST(RDFSimpleTests, StandardDeviation)
```
Comment (bluehood) [inline tree/dataframe/test/dataframe_simple.cxx]: you should use `TEST_P(RDFSimpleTests` instead of `TEST`, so the test will use the `RDFSimpleTests` fixture defined at the beginning of the file and will run in single and multi-thread.

File: tree/dataframe/src/RDFActionHelpers.cxx [hunk 10]
```diff
@@ -163,6 +163,62 @@ template void MeanHelper::Exec(unsigned int, const std::vector<char> &);
...
 template void MeanHelper::Exec(unsigned int, const std::vector<unsigned int> &);
 
+StdDevHelper::StdDevHelper(const std::shared_ptr<double> &meanVPtr, const unsigned int nSlots)
+   : fNSlots(nSlots), fResultStdDev(meanVPtr), fCounts(nSlots, 0), fMeans(nSlots, 0), fDistancesfromMean(nSlots, 0)
+{
+}
+
+void StdDevHelper::Exec(unsigned int slot, double v)
+{
+   // Online algorithm
```
Comment (bluehood) [inline tree/dataframe/src/RDFActionHelpers.cxx]: please mention the name of the algorithm in a comment so people can verify the implementation if they want to

File: tree/dataframe/test/dataframe_simple.cxx [hunk 11]
```diff
@@ -513,6 +514,113 @@ TEST_P(RDFSimpleTests, BookCustomAction)
    EXPECT_EQ(*maxSlot, nWorkers-1);
 }
 
+class StdDevHelper {
```
Comment (bluehood) [inline tree/dataframe/test/dataframe_simple.cxx]: unfortunate name collision, this is also the name of the action helper!

File: tree/dataframe/test/dataframe_simple.cxx [hunk 12]
```diff
@@ -513,6 +514,113 @@ TEST_P(RDFSimpleTests, BookCustomAction)
...
+   auto stdDev = tdf.Define("vector",
+                            []() {
+                               std::vector<int> v(3);
+                               v[0] = 0;
+                               v[1] = 1;
+                               v[2] = 2;
+                               return v;
+                            })
+                    .StdDev<std::vector<int>>("vector");
+   EXPECT_NEAR(*stdDev, 0.86602540378444, 0.0000000000001);
```
Comment (bluehood) [inline tree/dataframe/test/dataframe_simple.cxx]: this _really_ can't be `EXPECT_DOUBLE_EQ`?

Comment (imaxoi) [reply tree/dataframe/test/dataframe_simple.cxx]: If I try it, I get this:
```
128: tree/dataframe/test/dataframe_simple.cxx:567: Failure
128:       Expected: *stdDev
128:       Which is: 2.4494897427831779
128: To be equal to: 2.4494897427832
128:       Which is: 2.4494897427832001
```

**Comment by bluehood** [comment]
@phsft-bot build
